File: enve/Services/Import/AudiobookFileGrouping.swift
import Foundation

enum AudiobookFileGrouping {
    nonisolated static let minimumStandaloneBookSize: Int64 = 64 * 1_024 * 1_024

    nonisolated static func groups<Element>(
        _ files: [Element],
        name: (Element) -> String,
        bookEvidence: (Element) -> Bool = { _ in false },
        forcedStandalone: (Element) -> Bool = { _ in false }
    ) -> [[Element]] {
        guard files.count > 1 else { return files.isEmpty ? [] : [files] }

        let sortedFiles = sorted(files, name: name)
        let hasFilenameChapterEvidence = sortedFiles.contains {
            chapterGroupingKey(name($0)) != nil
        }
        let selfContainedExtensions: Set<String> = ["m4b", "m4a", "mp4"]
        let standaloneBooks = sortedFiles.filter {
            let filename = name($0)
            let hasChapterEvidence = chapterGroupingKey(filename) != nil
            return forcedStandalone($0)
                || isExplicitBook(filename)
                || (!hasChapterEvidence
                    && selfContainedExtensions.contains((filename as NSString).pathExtension.lowercased()))
                || (!hasFilenameChapterEvidence && bookEvidence($0))
        }
        let standaloneIds = Set(standaloneBooks.map { name($0).lowercased() })
        let candidates = sortedFiles.filter {
            !standaloneIds.contains(name($0).lowercased())
        }

        var result = standaloneBooks.map { [$0] }
        guard !candidates.isEmpty else { return result }

        let explicitKeys = Set(candidates.compactMap { chapterGroupingKey(name($0)) })
        let defaultKey = explicitKeys.count == 1 ? explicitKeys.first! : "multipart:ambiguous"
        let grouped = Dictionary(grouping: candidates) { file in
            chapterGroupingKey(name(file)) ?? defaultKey
        }
        result.append(contentsOf: grouped.values.map { group in
            sorted(group, name: name)
        })
        return result.sorted {
            guard let lhs = $0.first, let rhs = $1.first else { return !$0.isEmpty }
            return name(lhs).localizedStandardCompare(name(rhs)) == .orderedAscending
        }
    }

    nonisolated static func sorted<Element>(
        _ files: [Element],
        name: (Element) -> String
    ) -> [Element] {
        files.sorted { lhs, rhs in
            let lhsPosition = sequencePosition(name(lhs))
            let rhsPosition = sequencePosition(name(rhs))
            if lhsPosition.category != rhsPosition.category {
                return lhsPosition.category < rhsPosition.category
            }
            if lhsPosition.number != rhsPosition.number {
                return lhsPosition.number < rhsPosition.number
            }
            return name(lhs).localizedStandardCompare(name(rhs)) == .orderedAscending
        }
    }

    nonisolated private static func chapterGroupingKey(_ filename: String) -> String? {
        if let prefix = chapterSequencePrefix(filename) {
            return prefix.isEmpty
                ? "multipart:chapter-sequence"
                : "multipart:chapter-sequence:\(prefix)"
        }
        if isLeadingTrackNumber(filename) {
            return "multipart:chapter-sequence"
        }
        return multipartStem(filename).map { "multipart:\($0)" }
    }

    nonisolated static func inferredTitle(for filename: String) -> String {
        let filenameStem = (filename as NSString).deletingPathExtension
        guard multipartStem(filename) != nil else { return filenameStem }

        let title = filenameStem.replacingOccurrences(
            of: #"(?i)[\s._-]+(?:(?:part|track|chapter|disc|cd)[\s._-]*)?\d+$"#,
            with: "",
            options: .regularExpression
        )
        return title.isEmpty ? filenameStem : title
    }

    nonisolated private static func multipartStem(_ filename: String) -> String? {
        var normalized = (filename as NSString).deletingPathExtension.lowercased()
        for word in ["unabridged", "audiobook", "audio book"] {
            normalized = normalized.replacingOccurrences(of: word, with: " ")
        }
        normalized = normalized.replacingOccurrences(
            of: #"[\W_]+"#,
            with: " ",
            options: .regularExpression
        )
        let tokens = normalized.split(separator: " ").map(String.init)
        let explicitPartMarkers = ["part", "track", "chapter", "disc", "cd"]
        let hasPartMarker = tokens.dropLast().last.map(explicitPartMarkers.contains) == true
        guard let number = tokens.last,
            number.allSatisfy(\.isNumber),
            number.count >= 2 || hasPartMarker
        else {
            return nil
        }
        let stem = tokens.dropLast().joined(separator: " ")
        return stem.isEmpty ? nil : stem
    }

    nonisolated private static func chapterSequencePrefix(_ filename: String) -> String? {
        let stem = (filename as NSString).deletingPathExtension
        guard let range = stem.range(
            of: #"\b(?:chapter|track|part|disc|cd)[\s._#-]*\d+\b"#,
            options: [.regularExpression, .caseInsensitive]
        ) else {
            return nil
        }
        return normalizedTitle(String(stem[..<range.lowerBound]))
    }

    nonisolated private static func isLeadingTrackNumber(_ filename: String) -> Bool {
        let stem = (filename as NSString).deletingPathExtension
        return stem.range(of: #"^\s*\d+\b"#, options: .regularExpression) != nil
    }

    nonisolated private static func isExplicitBook(_ filename: String) -> Bool {
        guard chapterSequencePrefix(filename) == nil, !isLeadingTrackNumber(filename) else {
            return false
        }
        let stem = (filename as NSString).deletingPathExtension
        return stem.range(
            of: #"\b(?:book|volume|vol)[\s._#-]*\d+\b"#,
            options: [.regularExpression, .caseInsensitive]
        ) != nil
    }

    nonisolated private static func sequencePosition(_ filename: String) -> (category: Int, number: Int) {
        let normalized = normalizedTitle(filename)
        if ["opening credits", "intro", "introduction", "prologue"].contains(normalized) {
            return (0, 0)
        }
        if let chapterNumber = firstMatch(
            in: (filename as NSString).deletingPathExtension,
            pattern: #"^\s*chapter[\s._-]*(\d+)\b"#
        ) {
            return (1, chapterNumber)
        }
        if ["credits", "closing credits", "end credits", "epilogue"].contains(normalized) {
            return (3, 0)
        }
        return (2, 0)
    }

    nonisolated private static func normalizedTitle(_ filename: String) -> String {
        (filename as NSString).deletingPathExtension
            .lowercased()
            .replacingOccurrences(of: #"[\W_]+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    nonisolated private static func firstMatch(in value: String, pattern: String) -> Int? {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
            let match = regex.firstMatch(
                in: value,
                range: NSRange(value.startIndex..., in: value)
            ),
            let range = Range(match.range(at: 1), in: value)
        else {
            return nil
        }
        return Int(value[range])
    }
}

File: android/core/src/main/java/com/enve/core/data/importing/AudiobookFileGrouping.kt
package com.enve.core.data.importing

object AudiobookFileGrouping {
    const val MINIMUM_STANDALONE_BOOK_SIZE_BYTES: Long = 64L * 1_024L * 1_024L

    fun <T> groups(
        files: List<T>,
        name: (T) -> String,
        sizeBytes: (T) -> Long = { 0L },
        forcedStandalone: (T) -> Boolean = { false },
    ): List<List<T>> {
        if (files.size <= 1) return files.map(::listOf)

        val sortedFiles = sorted(files, name)
        val hasFilenameChapterEvidence = sortedFiles.any { chapterGroupingKey(name(it)) != null }
        val standaloneBooks = sortedFiles.filter { file ->
            val fileName = name(file)
            val hasChapterEvidence = chapterGroupingKey(fileName) != null
            forcedStandalone(file) ||
                isExplicitBook(fileName) ||
                (!hasChapterEvidence && extension(fileName) in selfContainedExtensions) ||
                (!hasFilenameChapterEvidence && sizeBytes(file) >= MINIMUM_STANDALONE_BOOK_SIZE_BYTES)
        }
        val standaloneNames = standaloneBooks.mapTo(mutableSetOf()) { name(it).lowercase() }
        val candidates = sortedFiles.filterNot { name(it).lowercase() in standaloneNames }
        if (candidates.isEmpty()) return standaloneBooks.map(::listOf)

        val explicitKeys = candidates.mapNotNull { chapterGroupingKey(name(it)) }.toSet()
        val defaultKey = explicitKeys.singleOrNull() ?: "multipart:ambiguous"
        val grouped = candidates.groupBy { chapterGroupingKey(name(it)) ?: defaultKey }
        return (standaloneBooks.map(::listOf) + grouped.values.map { sorted(it, name) })
            .sortedBy { group -> group.firstOrNull()?.let(name)?.lowercase().orEmpty() }
    }

    fun <T> sorted(files: List<T>, name: (T) -> String): List<T> =
        files.sortedWith(
            compareBy<T> { sequencePosition(name(it)).category }
                .thenBy { sequencePosition(name(it)).number }
                .thenBy { stem(name(it)).lowercase() }
                .thenBy { extension(name(it)) },
        )

    fun inferredTitle(fileName: String): String {
        val stem = stem(fileName)
        if (multipartStem(fileName) == null) return stem
        return stem.replace(trailingSequenceRegex, "").ifBlank { stem }
    }

    private fun chapterGroupingKey(fileName: String): String? {
        chapterSequencePrefix(fileName)?.let { prefix ->
            return if (prefix.isEmpty()) "multipart:chapter-sequence" else "multipart:chapter-sequence:$prefix"
        }
        if (leadingNumberRegex.containsMatchIn(stem(fileName))) return "multipart:chapter-sequence"
        return multipartStem(fileName)?.let { "multipart:$it" }
    }

    private fun chapterSequencePrefix(fileName: String): String? {
        val stem = stem(fileName)
        val match = chapterSequenceRegex.find(stem) ?: return null
        return normalizeTitle(stem.substring(0, match.range.first))
    }

    private fun multipartStem(fileName: String): String? {
        var normalized = normalizeTitle(stem(fileName))
        listOf("unabridged", "audiobook", "audio book").forEach { word ->
            normalized = normalized.replace(word, " ")
        }
        val tokens = normalizeTitle(normalized).split(' ').filter(String::isNotBlank)
        val number = tokens.lastOrNull()?.takeIf { token -> token.all(Char::isDigit) } ?: return null
        val hasPartMarker = tokens.dropLast(1).lastOrNull() in partMarkers
        if (number.length < 2 && !hasPartMarker) return null
        return tokens.dropLast(1).joinToString(" ").ifBlank { null }
    }

    private fun isExplicitBook(fileName: String): Boolean {
        val stem = stem(fileName)
        return chapterSequencePrefix(fileName) == null &&
            !leadingNumberRegex.containsMatchIn(stem) &&
            explicitBookRegex.containsMatchIn(stem)
    }

    private fun sequencePosition(fileName: String): SequencePosition {
        val normalized = normalizeTitle(stem(fileName))
        if (normalized in openingTitles) return SequencePosition(0, 0)
        val numbered = chapterSequenceRegex.find(stem(fileName))?.groupValues?.get(1)?.toIntOrNull()
            ?: leadingNumberCaptureRegex.find(stem(fileName))?.groupValues?.get(1)?.toIntOrNull()
        if (numbered != null) return SequencePosition(1, numbered)
        if (normalized in closingTitles) return SequencePosition(3, 0)
        return SequencePosition(2, 0)
    }

    private fun normalizeTitle(value: String): String =
        value.lowercase().replace(nonWordRegex, " ").trim()

    private fun extension(fileName: String): String = fileName.substringAfterLast('.', "").lowercase()

    private fun stem(fileName: String): String = fileName.substringBeforeLast('.', fileName)

    private data class SequencePosition(val category: Int, val number: Int)

    private val selfContainedExtensions = setOf("m4b", "m4a", "mp4")
    private val partMarkers = setOf("part", "track", "chapter", "disc", "cd")
    private val openingTitles = setOf("opening credits", "intro", "introduction", "prologue")
    private val closingTitles = setOf("credits", "closing credits", "end credits", "epilogue")
    private val explicitBookRegex = Regex("""\b(?:book|volume|vol)[\s._#-]*\d+\b""", RegexOption.IGNORE_CASE)
    private val chapterSequenceRegex = Regex("""\b(?:chapter|track|part|disc|cd)[\s._#-]*(\d+)\b""", RegexOption.IGNORE_CASE)
    private val leadingNumberRegex = Regex("""^\s*\d+\b""")
    private val leadingNumberCaptureRegex = Regex("""^\s*(\d+)\b""")
    private val trailingSequenceRegex = Regex("""(?i)[\s._-]+(?:(?:part|track|chapter|disc|cd)[\s._-]*)?\d+$""")
    private val nonWordRegex = Regex("""[\W_]+""")
}
